server/audit_math/raire.py

```diff
@@ -0,0 +1,318 @@
+from typing import Callable, Dict, Optional, List
+
+import numpy as np
+
+from .sampler_contest import Contest, CVRS
+from .raire_utils import (
+    NEBAssertion,
+    RaireAssertion,
+    RaireFrontier,
+    RaireNode,
+    find_best_audit,
+    perform_dive,
+)
+
+
+NEBMatrix = Dict[str, Dict[str, Optional[NEBAssertion]]]
+
+
+def make_neb_matrix(contest: Contest, cvrs: CVRS, asn_func) -> NEBMatrix:
+    """
+    Builds the NEB matrix for use by find_best_audit.
+
+    Input:
+        contest     - the contest being audited
+        cvrs        - a list of CVRS to be audited
+        asn_func    - the asn function to find assertion difficulty
+    Output:
+        neb_matrix  - a dict of dicts mapping candidate pairs to assertions
+    """
+    nebs: NEBMatrix = {
+        c: {d: None for d in contest.candidates} for c in contest.candidates
+    }
+
+    for cand in contest.candidates:
+        for other in contest.candidates:
+            if cand == other:
+                continue
+
+            asrn: NEBAssertion = NEBAssertion(contest.name, cand, other)
+
+            tally_cand: int = 0
+            tally_other: int = 0
+            for _, cvr in cvrs.items():
+                assert cvr  # for type checker
```

Comment:
so this typing error is solely a product of the `CVRS` data type in `sampler_contest` wrapping the `CVR` in `Optional`. Maybe I should write  a new type for this, since by the time we get here with audit results we will have thrown out all the empty ones.

Reply:
Will this code be run on both CVR data and audit results? I kind of thought it would just be CVR data since we compute the assertions before the audit when generating the sample size options.

I think an easy fix here is just to change this line to `assert cvr is not None`

Reply:
hmm, I suppose it probably will only be run on the CVR file and not on the audited data.

Reply:
My worry about `is not None` is that the linter will complain that that is redundant